File: problems/P-268-missing-number/solution2.cpp
/*
 * LeetCode Problem: 268. Missing Number
 * Difficulty: Easy
 * Topics: Array, Hash Table, Bit Manipulation, Math, Sorting, Binary Search
 * Date Solved: 05/07/2025
*/

#include <bits/stdc++.h>
using namespace std;

class Solution2 {
public:
  int missingNumber(vector<int>& nums) {
    sort(nums.begin(), nums.end());

    int left = 0, right = nums.size();
    while (left < right) {
      int mid = left + (right - left) / 2;
      if (nums[mid] > mid)
        right = mid;
      else
        left = mid + 1;
    }

    return left;
  }
};

File: problems/P-268-missing-number/solution2.java
/*
 * LeetCode Problem: 268. Missing Number
 * Difficulty: Easy
 * Topics: Array, Hash Table, Bit Manipulation, Math, Sorting, Binary Search
 * Date Solved: 05/07/2025
*/

public class solution2 {
  public int missingNumber(int[] nums) {
    int n = nums.length;
    int xor1 = 0, xor2 = 0;

    for (int i = 0; i < n; i++) {
      xor2 = xor2 ^ nums[i];
      xor1 = xor1 ^ (i + 1);
    }
    xor1 = xor1 ^ n;

    return (xor1 ^ xor2);
  }
}
